citesphere-model/src/main/java/edu/asu/diging/citesphere/data/bib/impl/CitationVersionDao.java

```diff
@@ -0,0 +1,76 @@
+package edu.asu.diging.citesphere.data.bib.impl;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.javers.core.Javers;
+import org.javers.core.metamodel.object.CdoSnapshot;
+import org.javers.core.metamodel.object.SnapshotType;
+import org.javers.repository.jql.QueryBuilder;
+import org.javers.shadow.Shadow;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+import edu.asu.diging.citesphere.data.bib.ICitationVersionsDao;
+import edu.asu.diging.citesphere.model.bib.ICitation;
+import edu.asu.diging.citesphere.model.bib.impl.Citation;
+import edu.asu.diging.citesphere.model.bib.impl.CitationVersion;
+
+@Repository
+public class CitationVersionDao implements ICitationVersionsDao {
+
+    @Autowired
+    private Javers javers;
+    
+    private static final String GROUP_PROPERTY = "group";
+
+    @Override
+    public List<CitationVersion> getVersions(String groupId, String key, int page, int pageSize) {
+        QueryBuilder jqlQuery = QueryBuilder.byInstanceId(key, Citation.class);
+        List<CdoSnapshot> versions = javers.findSnapshots(jqlQuery.build()).stream()
+                .filter(snapshot -> snapshot.getType() != SnapshotType.TERMINAL
+                        && groupId.equals((String) snapshot.getPropertyValue(GROUP_PROPERTY)))
+                .collect(Collectors.toList());
+        int offset = page * pageSize;
+        if (offset >= versions.size()) {
+            return new ArrayList<>();
+        }
+        List<CitationVersion> itemVersions = versions.subList(offset, Math.min(offset + pageSize, versions.size()))
+                .stream().map(snapshot -> toCitationVersion(snapshot, key)).collect(Collectors.toList());
+        return itemVersions;
+    }
+
+    @Override
+    public int getTotalCount(String groupId, String key) {
+        QueryBuilder jqlQuery = QueryBuilder.byInstanceId(key, Citation.class);
+        return javers.findSnapshots(jqlQuery.build()).stream()
+                .filter(snapshot -> snapshot.getType() != SnapshotType.TERMINAL
+                        && groupId.equals((String) snapshot.getPropertyValue(GROUP_PROPERTY)))
+                .collect(Collectors.toList()).size();
```

Comment:
same here. don't load everything into memory just get a count. I'm sure jql queries can return a count.

Reply:
It does not have count() functionality. Also, we need to check that the groupId matches, which again is not possible through query.

Reply:
what about `withCommitProperty`? Doesn't that allow you to filter the results? At least that would bring down the object count before loading it in memory.

Reply:
`withCommitProperty` refers to the properties of the commit metadata. It's empty as of now. I had used it earlier to manually add the citation key and then use `withCommitProperty` to extract all the versions. I can add the group Id which would be better than the current logic.